File: review-api/app/run_state.py
"""Canonical run → UI state mapping (Python twin of web/src/lib/runState.ts)."""
from __future__ import annotations

from typing import Any, Optional

from app.pipeline import (
    ARTIFACT_TYPE_TO_NODE,
    EXPLICIT_RUNNING_STAGES,
    HUMAN_GATES,
    PIPELINE_NODE_INDEX,
    PIPELINE_NODES,
)


def _parse_gate_suffix(stage: str, suffix: str) -> Optional[str]:
    for gate in HUMAN_GATES:
        if stage == f"{gate}{suffix}":
            return gate
    return None


def _nodes_from_artifacts(artifacts: list[dict[str, Any]]) -> set[str]:
    present: set[str] = set()
    for a in artifacts:
        node = ARTIFACT_TYPE_TO_NODE.get(a.get("type", ""))
        if node:
            present.add(node)
    return present


def _latest_decision(reviews: list[dict[str, Any]], gate: str) -> Optional[str]:
    found: Optional[str] = None
    for r in reviews:
        if r.get("gate") == gate:
            found = r.get("decision")
    return found


def _approved_gates(reviews: list[dict[str, Any]]) -> set[str]:
    return {g for g in HUMAN_GATES if _latest_decision(reviews, g) == "approved"}


def resolve_current_node_id(
    *,
    status: str,
    current_stage: Optional[str],
    pending_gate: Optional[str] = None,
    artifacts: Optional[list[dict[str, Any]]] = None,
    reviews: Optional[list[dict[str, Any]]] = None,
) -> str:
    stage = current_stage or ""
    artifacts = artifacts or []
    reviews = reviews or []

    rejected = _parse_gate_suffix(stage, "_rejected")
    if rejected:
        return rejected

    revising = _parse_gate_suffix(stage, "_revising")
    if revising:
        return revising

    if pending_gate in HUMAN_GATES:
        return pending_gate

    pending_stage = _parse_gate_suffix(stage, "_pending")
    if pending_stage:
        return pending_stage

    if status == "completed" or stage == "S11":
        return "S11"

    if stage in EXPLICIT_RUNNING_STAGES:
        return stage

    present = _nodes_from_artifacts(artifacts)
    approved = _approved_gates(reviews)
    for node in PIPELINE_NODES:
        nid = node["id"]
        if node["kind"] == "human":
            if nid not in approved:
                return nid
            continue
        if nid not in present:
            return nid
    return "S11"


def _list_pill(ui_status: str) -> dict[str, str]:
    mapping = {
        "awaiting_review": ("awaiting_review", "Paused – Awaiting Review"),
        "revising": ("in_progress", "In Progress"),
        "running": ("in_progress", "In Progress"),
        "completed": ("completed", "Completed"),
        "rejected": ("rejected", "Rejected"),
        "cancelled": ("cancelled", "Cancelled"),
        "failed": ("failed", "Failed"),
    }
    key, label = mapping.get(ui_status, ("failed", "Failed"))
    return {"key": key, "label": label}


def _build_pipeline_nodes(current_id: str, ui_status: str, reviews: list[dict[str, Any]]) -> list[dict[str, Any]]:
    current_index = PIPELINE_NODE_INDEX[current_id]
    approved = _approved_gates(reviews)
    nodes: list[dict[str, Any]] = []
    for index, defn in enumerate(PIPELINE_NODES):
        nid = defn["id"]
        if ui_status == "rejected" and index == current_index:
            state = "failed"
        elif ui_status == "completed" or (ui_status != "failed" and index < current_index):
            if defn["kind"] == "human" and (nid in approved or index < current_index):
                state = "approved"
            else:
                state = "completed"
        elif index == current_index:
            state = "failed" if ui_status == "failed" and defn["kind"] != "human" else "current"
        else:
            state = "pending"
        nodes.append(
            {
                "id": nid,
                "label": defn["label"],
                "shortLabel": defn["short_label"],
                "kind": defn["kind"],
                "state": state,
            }
        )
    return nodes


def derive_run_ui_state(
    *,
    status: str,
    current_stage: Optional[str],
    pending_gate: Optional[str] = None,
    artifacts: Optional[list[dict[str, Any]]] = None,
    reviews: Optional[list[dict[str, Any]]] = None,
) -> dict[str, Any]:
    artifacts = artifacts or []
    reviews = reviews or []
    stage = current_stage or ""
    current_node_id = resolve_current_node_id(
        status=status,
        current_stage=current_stage,
        pending_gate=pending_gate,
        artifacts=artifacts,
        reviews=reviews,
    )
    node_def = PIPELINE_NODES[PIPELINE_NODE_INDEX[current_node_id]]

    rejected_gate = _parse_gate_suffix(stage, "_rejected")
    revising_gate = _parse_gate_suffix(stage, "_revising")
    pending_from_stage = _parse_gate_suffix(stage, "_pending")
    pending = pending_gate if pending_gate in HUMAN_GATES else pending_from_stage

    action_gate: Optional[str] = None
    action_required = False
    run_health = "on_track"
    primary_cta = "none"

    if status == "cancelled":
        ui_status = "cancelled"
        run_health = "failed"
    elif rejected_gate or (status == "failed" and current_node_id in HUMAN_GATES):
        ui_status = "rejected"
        run_health = "failed"
        action_gate = rejected_gate or (current_node_id if current_node_id in HUMAN_GATES else None)
    elif status == "failed":
        ui_status = "failed"
        run_health = "failed"
    elif revising_gate:
        ui_status = "revising"
        action_gate = revising_gate
    elif pending and (status == "paused" or pending_from_stage or pending_gate):
        ui_status = "awaiting_review"
        action_required = True
        action_gate = pending
        run_health = "blocked"
        primary_cta = "open_review"
    elif status == "completed":
        ui_status = "completed"
        primary_cta = "view_report"
    else:
        ui_status = "running"

    return {
        "uiStatus": ui_status,
        "currentNodeId": current_node_id,
        "currentLabel": node_def["label"],
        "actionRequired": action_required,
        "actionGate": action_gate,
        "runHealth": run_health,
        "primaryCta": primary_cta,
        "stopRunAvailable": False,
        "shareAction": "copy_url",
        "pipelineNodes": _build_pipeline_nodes(current_node_id, ui_status, reviews),
        "listStatusPill": _list_pill(ui_status),
        "progressMessage": _progress_message(ui_status, stage),
    }


_PROGRESS_BY_STAGE = {
    "S1": "Normalizing the requirement…",
    "S2": "Analyzing ambiguities…",
    "S3": "Generating test cases…",
    "S4": "Generating test data…",
    "S5": "Generating automation…",
    "Compile": "Compiling Playwright…",
    "S6": "Running tests…",
    "S7": "Classifying failures…",
    "S8": "Scanning boundaries…",
    "S9": "Building the report…",
    "S10": "Writing the release summary…",
}


def _progress_message(ui_status: str, stage: str) -> Optional[str]:
    if ui_status == "revising" and _parse_gate_suffix(stage, "_revising"):
        return "Revising the prior phase, then this gate will reopen."
    if ui_status == "running":
        return _PROGRESS_BY_STAGE.get(stage)
    return None

File: web/src/lib/runState.ts
import {
  ARTIFACT_TYPE_TO_NODE,
  getNodeDef,
  HUMAN_GATES,
  isHumanGate,
  PIPELINE_NODE_INDEX,
  PIPELINE_NODES,
  type HumanGateId,
  type PipelineNodeId,
  type PipelineNodeKind,
} from "./pipeline";

export type DbRunStatus = "pending" | "running" | "paused" | "completed" | "failed" | "cancelled";

export type UiStatus =
  | "running"
  | "awaiting_review"
  | "revising"
  | "completed"
  | "failed"
  | "rejected"
  | "cancelled";

export type PipelineNodeState = "completed" | "current" | "pending" | "failed" | "approved";

export type RunHealth = "on_track" | "blocked" | "failed";

export type PrimaryCta = "open_review" | "view_report" | "none";

export type ListStatusPillKey =
  | "in_progress"
  | "awaiting_review"
  | "completed"
  | "failed"
  | "rejected"
  | "cancelled";

export interface ArtifactRef {
  type: string;
  version?: number;
}

export interface ReviewRef {
  gate: HumanGateId | string;
  decision: "pending" | "approved" | "rejected" | "changes_requested" | string;
}

export interface DeriveRunUiStateInput {
  status: DbRunStatus | string;
  current_stage: string | null | undefined;
  pending_gate?: HumanGateId | string | null;
  artifacts?: ArtifactRef[];
  reviews?: ReviewRef[];
}

export interface PipelineNodeUi {
  id: PipelineNodeId;
  label: string;
  shortLabel: string;
  kind: PipelineNodeKind;
  state: PipelineNodeState;
}

export interface ListStatusPill {
  key: ListStatusPillKey;
  label: string;
}

export interface RunUiState {
  uiStatus: UiStatus;
  currentNodeId: PipelineNodeId;
  currentLabel: string;
  actionRequired: boolean;
  actionGate: HumanGateId | null;
  runHealth: RunHealth;
  primaryCta: PrimaryCta;
  /** v1: Orchestrator has no cancel API — never show an active Stop. */
  stopRunAvailable: false;
  shareAction: "copy_url";
  pipelineNodes: PipelineNodeUi[];
  listStatusPill: ListStatusPill;
  /** Server-derived working copy; null when paused/terminal. */
  progressMessage: string | null;
}

const EXPLICIT_RUNNING_STAGES = new Set<PipelineNodeId>([
  "S1",
  "S2",
  "S3",
  "S4",
  "S5",
  "Compile",
  "S6",
  "S7",
  "S8",
  "S9",
  "S10",
  "S11",
]);

const PROGRESS_BY_STAGE: Partial<Record<PipelineNodeId, string>> = {
  S1: "Normalizing the requirement…",
  S2: "Analyzing ambiguities…",
  S3: "Generating test cases…",
  S4: "Generating test data…",
  S5: "Generating automation…",
  Compile: "Compiling Playwright…",
  S6: "Running tests…",
  S7: "Classifying failures…",
  S8: "Scanning boundaries…",
  S9: "Building the report…",
  S10: "Writing the release summary…",
};

export type PollFreshness = "live" | "updating" | "delayed";

/** Live = recent successful GET, not merely that a poll timer exists. */
export function pollFreshness(opts: {
  lastSuccessAt: number | null;
  consecutiveFailures: number;
  now?: number;
}): PollFreshness {
  if (opts.consecutiveFailures >= 2) return "delayed";
  if (opts.lastSuccessAt == null) return "updating";
  const age = (opts.now ?? Date.now()) - opts.lastSuccessAt;
  if (age < 5_000) return "live";
  if (age <= 15_000) return "updating";
  return "delayed";
}

export function formatUpdatedAgo(lastSuccessAt: number, now = Date.now()): string {
  const secs = Math.max(0, Math.floor((now - lastSuccessAt) / 1000));
  return secs === 0 ? "Updated just now" : `Updated ${secs}s ago`;
}

function parseGateSuffix(stage: string, suffix: "_pending" | "_revising" | "_rejected"): HumanGateId | null {
  for (const gate of HUMAN_GATES) {
    if (stage === `${gate}${suffix}`) return gate;
  }
  return null;
}

function nodesPresentFromArtifacts(artifacts: ArtifactRef[]): Set<PipelineNodeId> {
  const present = new Set<PipelineNodeId>();
  for (const a of artifacts) {
    const node = ARTIFACT_TYPE_TO_NODE[a.type];
    if (node) present.add(node);
  }
  return present;
}

function latestDecisionForGate(reviews: ReviewRef[], gate: HumanGateId): string | null {
  // Prefer the last matching review in array order (API returns chronological ASC).
  let found: string | null = null;
  for (const r of reviews) {
    if (r.gate === gate) found = r.decision;
  }
  return found;
}

function approvedGates(reviews: ReviewRef[]): Set<HumanGateId> {
  const set = new Set<HumanGateId>();
  for (const gate of HUMAN_GATES) {
    if (latestDecisionForGate(reviews, gate) === "approved") set.add(gate);
  }
  return set;
}

/**
 * Resolve which pipeline node is the frontier (current focus).
 */
export function resolveCurrentNodeId(input: DeriveRunUiStateInput): PipelineNodeId {
  const stage = input.current_stage ?? "";
  const pending = input.pending_gate;

  const rejected = parseGateSuffix(stage, "_rejected");
  if (rejected) return rejected;

  const revising = parseGateSuffix(stage, "_revising");
  if (revising) return revising;

  if (pending && isHumanGate(String(pending))) return pending as HumanGateId;

  const pendingStage = parseGateSuffix(stage, "_pending");
  if (pendingStage) return pendingStage;

  if (input.status === "completed" || stage === "S11") return "S11";

  if (EXPLICIT_RUNNING_STAGES.has(stage as PipelineNodeId)) {
    return stage as PipelineNodeId;
  }

  // Infer from artifacts: first node without evidence / approval.
  const present = nodesPresentFromArtifacts(input.artifacts ?? []);
  const approved = approvedGates(input.reviews ?? []);

  for (const node of PIPELINE_NODES) {
    if (node.kind === "human") {
      if (!approved.has(node.id as HumanGateId)) return node.id;
      continue;
    }
    if (!present.has(node.id)) return node.id;
  }

  return "S11";
}

function buildPipelineNodes(
  currentId: PipelineNodeId,
  uiStatus: UiStatus,
  reviews: ReviewRef[],
): PipelineNodeUi[] {
  const currentIndex = PIPELINE_NODE_INDEX[currentId];
  const approved = approvedGates(reviews);

  return PIPELINE_NODES.map((def, index) => {
    let state: PipelineNodeState;

    if (uiStatus === "rejected" && index === currentIndex) {
      state = "failed";
    } else if (uiStatus === "completed" || (uiStatus !== "failed" && index < currentIndex)) {
      if (def.kind === "human" && approved.has(def.id as HumanGateId)) {
        state = "approved";
      } else if (def.kind === "human" && index < currentIndex) {
        // Gate passed even if history sparse (e.g. inferred)
        state = "approved";
      } else {
        state = "completed";
      }
    } else if (index === currentIndex) {
      state = uiStatus === "failed" && def.kind !== "human" ? "failed" : "current";
    } else {
      state = "pending";
    }

    return {
      id: def.id,
      label: def.label,
      shortLabel: def.shortLabel,
      kind: def.kind,
      state,
    };
  });
}

function listPill(uiStatus: UiStatus): ListStatusPill {
  switch (uiStatus) {
    case "awaiting_review":
      return { key: "awaiting_review", label: "Paused – Awaiting Review" };
    case "revising":
    case "running":
      return { key: "in_progress", label: "In Progress" };
    case "completed":
      return { key: "completed", label: "Completed" };
    case "rejected":
      return { key: "rejected", label: "Rejected" };
    case "cancelled":
      return { key: "cancelled", label: "Cancelled" };
    case "failed":
    default:
      return { key: "failed", label: "Failed" };
  }
}

/**
 * Canonical run → UI state mapping. Use this everywhere (Run Detail, lists, dashboard).
 */
export function deriveRunUiState(input: DeriveRunUiStateInput): RunUiState {
  const stage = input.current_stage ?? "";
  const reviews = input.reviews ?? [];
  const currentNodeId = resolveCurrentNodeId(input);
  const nodeDef = getNodeDef(currentNodeId);

  let uiStatus: UiStatus;
  let actionGate: HumanGateId | null = null;
  let actionRequired = false;
  let runHealth: RunHealth = "on_track";
  let primaryCta: PrimaryCta = "none";

  const rejectedGate = parseGateSuffix(stage, "_rejected");
  const revisingGate = parseGateSuffix(stage, "_revising");
  const pendingFromStage = parseGateSuffix(stage, "_pending");
  const pendingGate =
    (input.pending_gate && isHumanGate(String(input.pending_gate))
      ? (input.pending_gate as HumanGateId)
      : null) ?? pendingFromStage;

  if (input.status === "cancelled") {
    uiStatus = "cancelled";
    runHealth = "failed";
  } else if (rejectedGate || (input.status === "failed" && isHumanGate(currentNodeId))) {
    uiStatus = "rejected";
    runHealth = "failed";
    actionGate = rejectedGate ?? (isHumanGate(currentNodeId) ? currentNodeId : null);
  } else if (input.status === "failed") {
    uiStatus = "failed";
    runHealth = "failed";
  } else if (revisingGate) {
    uiStatus = "revising";
    actionGate = revisingGate;
    runHealth = "on_track";
  } else if (pendingGate && (input.status === "paused" || pendingFromStage || input.pending_gate)) {
    uiStatus = "awaiting_review";
    actionRequired = true;
    actionGate = pendingGate;
    runHealth = "blocked";
    primaryCta = "open_review";
  } else if (input.status === "completed") {
    uiStatus = "completed";
    primaryCta = "view_report";
    runHealth = "on_track";
  } else {
    uiStatus = "running";
    runHealth = "on_track";
  }

  const pipelineNodes = buildPipelineNodes(currentNodeId, uiStatus, reviews);

  return {
    uiStatus,
    currentNodeId,
    currentLabel: nodeDef.label,
    actionRequired,
    actionGate,
    runHealth,
    primaryCta,
    stopRunAvailable: false,
    shareAction: "copy_url",
    pipelineNodes,
    listStatusPill: listPill(uiStatus),
    progressMessage: progressMessageFor(uiStatus, stage),
  };
}

function progressMessageFor(uiStatus: UiStatus, stage: string): string | null {
  if (uiStatus === "revising" && parseGateSuffix(stage, "_revising")) {
    return "Revising the prior phase, then this gate will reopen.";
  }
  if (uiStatus === "running") {
    return PROGRESS_BY_STAGE[stage as PipelineNodeId] ?? null;
  }
  return null;
}
